assets/js/blocks-registry/payment-methods/registry.js

```diff
@@ -0,0 +1,33 @@
+/**
+ * Internal dependencies
+ */
+import {
+	assertValidPaymentMethod,
+	assertValidPaymentMethodComponent,
+} from './assertions';
+
+// currently much leeway is given to the payment method for the shape of their components. We should investigate payment methods
```

Comment:
Is it possible that a payment form might require additional details, not based on the payment method, but based on the items in the cart? Maybe not in this case, but it's been a major design consideration for CompositeCheckout which guided a lot of the abstractions used and I figured it's worth asking about here.

Reply:
yes, it will require additional details.  That's why when the payment method component is rendered it's [fed all these extra details as props](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1349/files#diff-cea364bc272eb722c3373b4d6e828d14R69-R73) via what is available in the `checkoutData` object (line items, totals, fields, whatever we decide are common things a payment method might need to know).

Reply:
Cool. That was the initial approach I took in composite checkout as well, but a problem we discovered is that it makes the payment method responsible for knowing how to interpret that data for all use cases. Again, it's very possible this isn't an issue for Woo, but in our case, each parent component might want to display the form in a totally different manner; passing in only the data means that our component needs to understand what that data means, so if we add a new kind of requirement, we'd have to change the _component_ rather than just the parent.

For us, the solution was to make the contact information fields a React element that's passed into the component as one of its props, and to have the default fields be an additional export of the package.

Reply:
This might prove problematic with the approach we're trying to make, since form fields will be blocks or atomic blocks that use can reorder and interact with, I think we can expose hooks and APIs for extensions to use add extra fields

Reply:
> This might prove problematic with the approach we're trying to make, since form fields will be blocks or atomic blocks that use can reorder and interact with, I think we can expose hooks and APIs for extensions to use add extra fields

- I think we'll still have to protect certain required fields. 
- I could see potentially allowing re-ordering of steps but that won't affect the availability of the data.
- I think billing fields will need to be controlled by the payment method (not present in this pull currently but would work similarly to the other existing api) and protected. 
- Extensions could add extra fields but the payment method should only be concerned about things affecting line items and totals so I'm not sure how much that will affect things there because from the perspective of payment methods, as long as we're providing the data payment methods need to complete a transaction (totals, currency, line items, potential subscription recurrence) then it doesn't really matter what extra data exists in the form.  The payment methods themselves might need extensibility for things like subscription but imo that's another reason why payment methods should be their own modules so that they have control over how payments are processed and how that connects with anything else they may extend in woo.

Reply:
👍 I should have been more specific. In composite checkout we actually have two smaller forms that are part of the checkout: the first is controlled by the payment method and has all the data it requires; the second (actually, an arbitrary number of additional ones as they are stored in "step objects") is controlled by the parent component and is dependent on the line items. It sounds like the sort of flexibility in the second form is just unnecessary here. Thanks for the explanation!